File: HelloGoodbyeUniv/AllYourBaseModel.m
//
//  AllYourBaseModel.m
//  AllYourBase
//
//  Created by Tiago Henriques on 3/29/11.
//  Copyright 2011 notknot. All rights reserved.
//

#import "AllYourBaseModel.h"

@implementation AllYourBaseModel

@synthesize currentDigits;
@synthesize previousDigits;
@synthesize currentOperation;
@synthesize previousOperation;
@synthesize previousExpression;
@synthesize resultDigits;
@synthesize secondaryDisplay;
@synthesize mainDisplay;
@synthesize error;
@synthesize previousFirstOperand;
@synthesize previousSecondOperand;
@synthesize base;

# pragma mark overridden methods
- (id)init
{
    self = [super init];
    if (self) {
        self.base = 10;
        self.previousFirstOperand = 0;
        self.previousSecondOperand = 0;
        self.error = nil;
        self.previousDigits = nil;
        self.currentDigits = [[[Digits alloc] initWithBase:10] autorelease];
        [self updateDisplays];
    }
    return self;
}

- (void)dealloc
{
    [self releaseMembers];
    [super dealloc];
}

# pragma mark UITabBarControllerDelegate optional methods

- (void)tabBarController:(UITabBarController *)tabBarController didSelectViewController:(UIViewController *)viewController
{
    NSLog(@"%@ selected", viewController);
}

# pragma mark instance methods

- (void)releaseMembers
{
    self.previousDigits = nil;
    self.currentDigits = nil;
    self.resultDigits = nil;
    self.secondaryDisplay = nil;
    self.mainDisplay = nil;
    self.currentOperation = nil;
    self.previousOperation = nil;
    self.previousExpression = nil;
    self.previousFirstOperand = 0;
    self.previousSecondOperand = 0;
    self.error = nil;
}

# pragma mark -

- (void)updateMainDisplay
{
    NSString *firstOperand = @"";
    NSString *paddedOperation = @"";
    NSString *secondOperand = @"";
    
    if (self.currentOperation) {
        firstOperand = self.previousDigits.description;
        paddedOperation = [NSString stringWithFormat:@" %@ ", self.currentOperation];
        secondOperand = self.currentDigits.description;  
    } else {
//        secondOperand = [self.currentDigits.description isEqualToString:@""] ? @"0" : self.currentDigits.description;
        secondOperand = self.currentDigits.description;
    }

    NSString *oldDisplay = self.mainDisplay;
    NSString *newDisplay = [NSString stringWithFormat:
                            @"%@%@%@%@"
                            , self.previousOperation ? @"= " : @""
                            , firstOperand ? firstOperand : @""
                            , paddedOperation
                            , secondOperand ? secondOperand : @""
                            ];
    if (![oldDisplay isEqualToString:newDisplay]) {
        self.mainDisplay = newDisplay;
    }
}

- (void)updateSecondaryDisplay
{
    if (self.error) {
        self.secondaryDisplay = [self.error localizedDescription];
        return; // early return if in erroneous state
    }
    
    NSString *oldDisplay = self.secondaryDisplay;
    NSString *newDisplay = [NSString stringWithFormat:
                            @"%@"
                            , self.previousOperation ? self.previousExpression : @""
                            ];
    
    if (![oldDisplay isEqualToString:newDisplay]) {
        self.secondaryDisplay = newDisplay;
    }
}

- (void)updateDisplays
{
    [self updateSecondaryDisplay];
    [self updateMainDisplay];
}

# pragma mark -

- (void)setBase:(int)someBase
{
    if ((someBase < 2) || (someBase > 100)) {
        NSLog(@"only bases from 2 to 100 are supported");
        return;
    }
    
    if (someBase == self.base) {
        NSLog(@"no sense in changing base to the same base");
        return;
    }
    
    base = someBase;
    
    if (self.currentDigits) {
        if (self.currentDigits.unsignedDigits) {
//            double currentValue = ((FloatingDigits *)self.currentDigits).doubleValue;
//            self.currentDigits = [[Digits alloc] initWithDouble:currentValue base:someBase];
            long long int currentValue = self.currentDigits.integerValue;
            self.currentDigits = [[Digits alloc] initWithLongLong:currentValue base:someBase];
        } else
        {
            self.currentDigits = [[Digits alloc] initWithBase:someBase];
        }
    }
    
    if (self.previousDigits) {
        if (self.previousDigits.unsignedDigits) {
//            double previousValue = ((FloatingDigits *)self.previousDigits).doubleValue;
//            self.previousDigits = [[Digits alloc] initWithDouble:previousValue base:someBase];
            long long int previousValue = self.previousDigits.integerValue;
            self.previousDigits = [[Digits alloc] initWithLongLong:previousValue base:someBase];
        } else
        {
            self.previousDigits = [[Digits alloc] initWithBase:someBase];
        }
    }
    
    if (self.previousOperation && (self.previousOperation != @"=")) {
        NSString *firstOperand = [Digits convertInteger:self.previousFirstOperand toBase:someBase];
        NSString *secondOperand = [Digits convertInteger:self.previousSecondOperand toBase:someBase];
        self.previousExpression = [NSString stringWithFormat:@"%@ %@ %@"
                                   , firstOperand ? firstOperand : @""
                                   , self.previousOperation ? self.previousOperation : @""
                                   , secondOperand ? secondOperand : @""
                                   ];
    } else
    {
        self.previousFirstOperand = 0;
        self.previousSecondOperand = 0;
        self.previousExpression = nil;
        self.previousOperation = nil;
    }

    [self updateDisplays];
}

# pragma mark -

- (BOOL)performPendingOperationWithError:(NSError **)operationError
{
    Digits *operationResultDigits = nil;
    
    if ([self.currentOperation isEqualToString:@"+"]) {
        operationResultDigits = [self.previousDigits plus:self.currentDigits withError:operationError];
    } else if ([self.currentOperation isEqualToString:@"-"]) {
        operationResultDigits = [self.previousDigits minus:self.currentDigits withError:operationError];
    } else if ([self.currentOperation isEqualToString:@"*"]) {
        operationResultDigits = [self.previousDigits times:self.currentDigits withError:operationError];
    } else if ([self.currentOperation isEqualToString:@"/"]) {
        operationResultDigits = [self.previousDigits divide:self.currentDigits withError:operationError];
    } else if ([self.currentOperation isEqualToString:@"^"]) {
        operationResultDigits = [self.previousDigits power:self.currentDigits withError:operationError];
    } else {
        NSLog(@"unknown operation '%@'", self.currentOperation);
        if (operationError) {            
            NSDictionary *userDict = [[NSDictionary dictionaryWithObjectsAndKeys:
                                       NSLocalizedString(@"unknown operation '%@'", self.currentOperation),
                                       NSLocalizedDescriptionKey,
                                       nil] retain];
            NSError *localError = [[[NSError alloc] initWithDomain:NSCocoaErrorDomain code:EPERM userInfo:userDict] autorelease];
            *operationError = localError;
            [userDict release];
        }
    }
    
    self.resultDigits = operationResultDigits;
    
    if (*operationError) {
        return NO;
    } else
    {
        return YES;
    }
}

# pragma mark -

- (void)resultPressed
{
    if (self.error) {
        NSLog(@"No action - result does nothing after error");
        return;
    }
    
    if (!self.currentDigits.unsignedDigits && self.currentOperation) {
        NSLog(@"No action - pending operation cannot be performed without a 2nd operand");
    } else if (!self.currentDigits.unsignedDigits && !self.currentOperation) {
        NSLog(@"No action - result cannot be performed on empty operand");
    } else if (self.currentDigits.unsignedDigits && self.currentOperation) {
        NSError *operationError = nil;
        [self performPendingOperationWithError:&operationError];
        
        if (!self.resultDigits && operationError) {
            NSLog(@"binary operation error after pressing result");
            self.error = operationError;
        } else {
            self.previousFirstOperand = [self.previousDigits integerValue];
            self.previousOperation = self.currentOperation;
            self.previousSecondOperand = [self.currentDigits integerValue];
            self.previousExpression = [NSString stringWithFormat:@"%@ %@ %@"
                                       , self.previousDigits ? self.previousDigits.description : @""
                                       , self.currentOperation ? self.currentOperation : @""
                                       , self.currentDigits ? self.currentDigits.description : @""
                                       ];
            self.previousDigits = nil;
            self.currentDigits = self.resultDigits;
            self.currentOperation = nil;        }
    } else if (self.currentDigits.unsignedDigits && !self.currentOperation) {
        self.previousOperation = @"=";
        self.previousExpression = self.currentDigits.description;
        self.resultDigits = [[Digits alloc] 
                             initWithLongLong:[self.currentDigits integerValue] 
                             base:self.base];
        self.currentDigits = self.resultDigits;
        self.previousDigits = nil;
        self.currentOperation = nil;
    } else {
//        assert(NO); // should never reach this point
    }
    [self updateDisplays];
}

- (void)binaryOperationPressed:(NSString *)operation 
{
    if (self.error) {
        NSLog(@"No action - operations do nothing after error");
        return;
    }
    
    if (!self.currentDigits.unsignedDigits && !self.currentOperation) {
        NSLog(@"No action - operations do nothing without a 2nd operand");
    } else if (!self.currentDigits.unsignedDigits && self.currentOperation) {
        if (!self.currentDigits.startsWithMinus) {
            self.currentOperation = operation;
            // pressing another operation when no 2nd operand digit or minus sign has been input cancels the pending operation
        } else {
            // pressing another operation when a minus sign has been input does nothing            
        }
    } else if (self.currentDigits.unsignedDigits && self.currentOperation) {
        NSError *operationError = nil;
        [self performPendingOperationWithError:&operationError];
        
        if (!self.resultDigits && operationError) {
            NSLog(@"binary operation error after chaining operation");
            self.error = operationError;
        } else {            
            self.previousFirstOperand = [self.previousDigits integerValue];
            self.previousOperation = self.currentOperation;
            self.previousSecondOperand = [self.currentDigits integerValue];
            self.previousExpression = [NSString stringWithFormat:@"%@ %@ %@"
                                       , self.previousDigits ? self.previousDigits.description : @""
                                       , self.currentOperation ? self.currentOperation : @""
                                       , self.currentDigits ? self.currentDigits.description : @""
                                       ];
            self.previousDigits = self.resultDigits;
            self.currentDigits = [[[Digits alloc] initWithBase:self.base] autorelease];
            self.currentOperation = operation;
        }
    } else { // no pending operation
        self.previousDigits = self.currentDigits;
        self.currentOperation = operation;
        self.currentDigits = [[[Digits alloc] initWithBase:self.base] autorelease];
    }
    
    [self updateDisplays];
}

- (void)digitPressed:(NSString *)digit
{
    if (self.error) {
        NSLog(@"No action - digits do nothing after error");
        return;
    }
    
    if (self.previousOperation) {
        self.currentDigits = [[[Digits alloc] initWithBase:self.base] autorelease];
    }
    
    [self.currentDigits pushDigit:digit];
    
    self.previousOperation = nil;
    [self updateDisplays];
}

- (void)deletePressed
{    
    if (self.previousOperation) {
        NSLog(@"No action - delete does not alter calculation results");
        return;
    }
    
    NSString *originalDigits = self.currentDigits.signedDigits;
    NSString *poppedDigit = [self.currentDigits popDigit];
    NSLog(@"popped digit '%@' from digits '%@'", poppedDigit, originalDigits);

    if (self.error) {
        self.error = nil;
        NSLog(@"cleaned error after pressing delete");
    }

    [self updateDisplays];    
}

- (void)negatePressed
{

    if (self.error) {
        NSLog(@"No action - negate does nothing after error");
        return;
    }
    
    if (self.previousOperation) {
        self.previousOperation = nil;
        self.previousExpression = nil;  
        self.previousFirstOperand = 0;
        self.previousSecondOperand = 0;
        self.resultDigits = nil;
        self.currentDigits = [[[Digits alloc] initWithBase:self.base] autorelease];
    }

    [self.currentDigits negate];
    [self updateDisplays];    
}

- (void)cleanPressed
{
    [self releaseMembers];
    self.currentDigits = [[[Digits alloc] initWithBase:self.base] autorelease];
    [self updateDisplays];
}

- (void)shiftLeftPressed
{
    
}

- (void)shiftRightPressed
{
    
}

- (void)percentPressed
{
    
}

- (void)EEPressed
{
    
}

@end

File: AllYourBase/AllYourBaseModel.swift
//
//  AllYourBaseModel.swift
//  AllYourBase
//
//  The calculator's state machine. Ported from the Objective-C
//  `AllYourBaseModel`, which used KVO on `mainDisplay`/`secondaryDisplay`;
//  this uses `@Published` / `ObservableObject` instead so SwiftUI views can
//  observe it directly.
//
//  A single model instance is shared across every base's tab (see
//  `ContentView`), so switching tabs re-renders the same in-progress
//  calculation in a different base, exactly like the original app.
//

import Foundation

final class AllYourBaseModel: ObservableObject {
    @Published private(set) var base: Int

    @Published var currentDigits: Digits
    @Published var previousDigits: Digits?
    @Published var resultDigits: Digits?
    @Published var currentOperation: String?
    @Published var previousOperation: String?
    @Published var previousExpression: String?
    @Published private(set) var mainDisplay: String = ""
    @Published private(set) var secondaryDisplay: String = ""
    @Published var error: DigitsError?

    private var previousFirstOperand: Int64 = 0
    private var previousSecondOperand: Int64 = 0

    init(base: Int = 10) {
        self.base = base
        self.currentDigits = Digits(base: base)!
        updateDisplays()
    }

    // MARK: Base switching

    func setBase(_ someBase: Int) {
        guard someBase >= Digits.minBase, someBase <= Digits.maxBase else {
            print("only bases from \(Digits.minBase) to \(Digits.maxBase) are supported")
            return
        }
        guard someBase != base else {
            return
        }

        base = someBase

        if currentDigits.unsignedDigits != nil {
            currentDigits = Digits(longLong: currentDigits.integerValue, base: someBase)!
        } else {
            currentDigits = Digits(base: someBase)!
        }

        if let previous = previousDigits {
            if previous.unsignedDigits != nil {
                previousDigits = Digits(longLong: previous.integerValue, base: someBase)!
            } else {
                previousDigits = Digits(base: someBase)!
            }
        }

        if let operation = previousOperation, operation != "=" {
            let firstOperand = Digits.convertInteger(previousFirstOperand, toBase: someBase)
            let secondOperand = Digits.convertInteger(previousSecondOperand, toBase: someBase)
            previousExpression = "\(firstOperand) \(operation) \(secondOperand)"
        } else {
            previousFirstOperand = 0
            previousSecondOperand = 0
            previousExpression = nil
            previousOperation = nil
        }

        updateDisplays()
    }

    // MARK: Display

    private func updateMainDisplay() {
        var firstOperand = ""
        var paddedOperation = ""
        let secondOperand = currentDigits.description

        if let operation = currentOperation {
            firstOperand = previousDigits?.description ?? ""
            paddedOperation = " \(operation) "
        }

        let prefix = previousOperation != nil ? "= " : ""
        mainDisplay = "\(prefix)\(firstOperand)\(paddedOperation)\(secondOperand)"
    }

    private func updateSecondaryDisplay() {
        if let error {
            secondaryDisplay = error.errorDescription ?? ""
            return
        }
        secondaryDisplay = previousOperation != nil ? (previousExpression ?? "") : ""
    }

    private func updateDisplays() {
        updateSecondaryDisplay()
        updateMainDisplay()
    }

    // MARK: Operations

    private func performPendingOperation() throws -> Digits {
        guard let previous = previousDigits, let operation = currentOperation else {
            throw DigitsError.overflow("unknown operation")
        }
        switch operation {
        case "+": return try previous.plus(currentDigits)
        case "-": return try previous.minus(currentDigits)
        case "*": return try previous.times(currentDigits)
        case "/": return try previous.divide(currentDigits)
        case "^": return try previous.power(currentDigits)
        default:
            print("unknown operation '\(operation)'")
            throw DigitsError.overflow("unknown operation")
        }
    }

    func resultPressed() {
        guard error == nil else {
            print("No action - result does nothing after error")
            return
        }

        let hasSecondOperand = currentDigits.unsignedDigits != nil

        if !hasSecondOperand && currentOperation != nil {
            print("No action - pending operation cannot be performed without a 2nd operand")
        } else if !hasSecondOperand && currentOperation == nil {
            print("No action - result cannot be performed on empty operand")
        } else if hasSecondOperand && currentOperation != nil {
            do {
                let result = try performPendingOperation()
                resultDigits = result
                previousFirstOperand = previousDigits?.integerValue ?? 0
                previousOperation = currentOperation
                previousSecondOperand = currentDigits.integerValue
                previousExpression = "\(previousDigits?.description ?? "") \(currentOperation ?? "") \(currentDigits.description)"
                previousDigits = nil
                currentDigits = result
                currentOperation = nil
            } catch let digitsError as DigitsError {
                print("binary operation error after pressing result")
                error = digitsError
            } catch {
                self.error = .overflow("unknown")
            }
        } else {
            previousOperation = "="
            previousExpression = currentDigits.description
            let result = Digits(longLong: currentDigits.integerValue, base: base)!
            resultDigits = result
            currentDigits = result
            previousDigits = nil
            currentOperation = nil
        }

        updateDisplays()
    }

    func binaryOperationPressed(_ operation: String) {
        guard error == nil else {
            print("No action - operations do nothing after error")
            return
        }

        let hasSecondOperand = currentDigits.unsignedDigits != nil

        if !hasSecondOperand && currentOperation == nil {
            print("No action - operations do nothing without a 2nd operand")
        } else if !hasSecondOperand && currentOperation != nil {
            // Pressing another operation cancels the pending one, unless a
            // lone minus sign has already been entered for the 2nd operand.
            if !currentDigits.startsWithMinus {
                currentOperation = operation
            }
        } else if hasSecondOperand && currentOperation != nil {
            do {
                let result = try performPendingOperation()
                resultDigits = result
                previousFirstOperand = previousDigits?.integerValue ?? 0
                previousOperation = currentOperation
                previousSecondOperand = currentDigits.integerValue
                previousExpression = "\(previousDigits?.description ?? "") \(currentOperation ?? "") \(currentDigits.description)"
                previousDigits = result
                currentDigits = Digits(base: base)!
                currentOperation = operation
            } catch let digitsError as DigitsError {
                print("binary operation error after chaining operation")
                error = digitsError
            } catch {
                self.error = .overflow("unknown")
            }
        } else {
            previousDigits = currentDigits
            currentOperation = operation
            currentDigits = Digits(base: base)!
        }

        updateDisplays()
    }

    func digitPressed(_ digit: String) {
        guard error == nil else {
            print("No action - digits do nothing after error")
            return
        }

        if previousOperation != nil {
            currentDigits = Digits(base: base)!
        }

        currentDigits.pushDigit(digit)
        previousOperation = nil
        updateDisplays()
    }

    func deletePressed() {
        guard previousOperation == nil else {
            print("No action - delete does not alter calculation results")
            return
        }

        currentDigits.popDigit()

        if error != nil {
            error = nil
            print("cleaned error after pressing delete")
        }

        updateDisplays()
    }

    func negatePressed() {
        guard error == nil else {
            print("No action - negate does nothing after error")
            return
        }

        if previousOperation != nil {
            previousOperation = nil
            previousExpression = nil
            previousFirstOperand = 0
            previousSecondOperand = 0
            resultDigits = nil
            currentDigits = Digits(base: base)!
        }

        currentDigits.negate()
        updateDisplays()
    }

    func cleanPressed() {
        previousDigits = nil
        resultDigits = nil
        currentOperation = nil
        previousOperation = nil
        previousExpression = nil
        previousFirstOperand = 0
        previousSecondOperand = 0
        error = nil
        currentDigits = Digits(base: base)!
        updateDisplays()
    }
}
